apps/src/applab/appStorage.js

```diff
@@ -2,13 +2,62 @@
 
 /* global Applab */
 
-var RecordListener = require('./RecordListener');
+var Firebase = require("firebase");
 
 /**
  * Namespace for app storage.
  */
 var AppStorage = module.exports;
 
+// HACKHACK: Need to make this configurable
+var BASE_DB_URL = 'https://radiant-fire-5518.firebaseio.com';
+
+var databaseCache = {};
+function getDatabase(channelId) {
+  var db = databaseCache[channelId];
+  if (db == null) {
+    db = new Firebase(BASE_DB_URL + '/v3/shared-tables/' + channelId);
+    databaseCache[channelId] = db;
+  }
+  console.log(db);
+  return db;
+}
+
+function getKeyValues(channelId) {
+  var kv = getDatabase(channelId).child('keys');
+  console.log(kv);
+  return kv;
+}
+
+function getTable(channelId, tableName) {
+  return getDatabase(channelId).child('tables').child(tableName);
+}
+
+/**
+ * Increments the given counter, specified by channelId and counterName.
+ * Calls onSuccess with the counter value, or onError with an error string
+ * if it was not possible to determine the value of the counter.
+ */
+function getCounter(channelId, counterName, onSuccess, onError) {
+  var counters = getDatabase(channelId).child('counters');
+  counters.child(counterName).transaction(function(currentValue) {
+    return (currentValue || 0) + 1;
+  }, function(err, committed, data) {
+    if(err) {
+      onError(err);
+    }
+    else if (committed) {
+      // This should always be available if there was no error, since our
+      // update functions always returns a value.
+      onSuccess(data.val());
+    } else {
+      onError("Unexpected error");
+    }
+  });
+}
+
+var firebase = new Firebase('https://radiant-fire-5518.firebaseio.com/applab');
```

Comment:
This is a developer instance of Firebase that has only data from a few
different experimental apps I wrote; is that your question?  We're not
paying anything for this right now.

On Mon, Mar 28, 2016 at 4:04 PM, David Bailey <notifications@github.com>
wrote:

> In apps/src/applab/appStorage.js
> <https://github.com/code-dot-org/code-dot-org/pull/7281#discussion_r57648552>
> :
>
> > +    return (currentValue || 0) + 1;
> > +  }, function(err, committed, data) {
> > +    if(err) {
> > +      onError(err);
> > +    }
> > +    else if (committed) {
> > +      // This should always be available if there was no error, since our
> > +      // update functions always returns a value.
> > +      onSuccess(data.val());
> > +    } else {
> > +      onError("Unexpected error");
> > +    }
> > +  });
> > +}
> > +
> > +var firebase = new Firebase('https://radiant-fire-5518.firebaseio.com/applab');
>
> @philbogle <https://github.com/philbogle> this instance of Firebase is
> currently unused, correct? Just making sure I'm understanding the prototype
> correctly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/7281/files/24db5f6c175eb93003ec2b2d537898cdb976b98d#r57648552>
>